File: syncloud/src/main/java/org/syncloud/android/tasks/ProgressAsyncTask.java
package org.syncloud.android.tasks;

import android.os.AsyncTask;

import org.syncloud.android.Progress;

public class ProgressAsyncTask<TParams, TResult> extends AsyncTask<TParams, Void, AsyncResult<TResult>> {

    public interface Success<TResult> {
        void run(TResult result);
    }

    public interface Completed<TResult> {
        void run(AsyncResult<TResult> result);
    }

    public interface Work<TParams, TResult> {
        AsyncResult<TResult> run(TParams... args);
    }

    private String title;
    private Progress progress;
    private Work<TParams, TResult> work;
    private String errorMessage;
    private Success<TResult> success;
    private Completed<TResult> completed;

    public ProgressAsyncTask(Progress progress, String title, Work<TParams, TResult> work) {
        this.progress = progress;
        this.title = title;
        this.work = work;
    }

    public ProgressAsyncTask() {
        this(null, null, null);
    }

    @Override
    protected AsyncResult<TResult> doInBackground(TParams... args) {
        if (work != null) {
            try {
                return work.run(args);
            } catch (Throwable th) {
                return AsyncResult.error(th.getMessage());
            }
        } else {
            return null;
        }
    }

    public ProgressAsyncTask<TParams, TResult> setErrorMessage(String message) {
        this.errorMessage = message;
        return this;
    }

    public ProgressAsyncTask<TParams, TResult> setTitle(String title) {
        this.title = title;
        return this;
    }

    public ProgressAsyncTask<TParams, TResult> setProgress(Progress progress) {
        this.progress = progress;
        return this;
    }

    public ProgressAsyncTask<TParams, TResult> onSuccess(Success<TResult> success) {
        this.success = success;
        return this;
    }

    public ProgressAsyncTask<TParams, TResult> onCompleted(Completed<TResult> completed) {
        this.completed = completed;
        return this;
    }

    public ProgressAsyncTask<TParams, TResult> doWork(Work<TParams, TResult> work) {
        this.work = work;
        return this;
    }

    @Override
    protected void onPreExecute() {
        if (progress != null) {
            progress.start();
            if (title != null)
                progress.title(title);
        }
    }

    @Override
    protected void onPostExecute(AsyncResult<TResult> result) {
        if (progress != null) {
            if (result != null && !result.hasValue() && errorMessage != null)
                progress.error(errorMessage);
            else
                progress.stop();
        }
        if (result == null)
            success.run(null);
        else
            if (result.hasValue() && success != null)
                success.run(result.getValue());

        if (completed != null)
            completed.run(result);
    }
}

File: syncloud/src/main/java/org/syncloud/android/tasks/ProgressAsyncTask.kt
package org.syncloud.android.tasks

import android.os.AsyncTask
import org.syncloud.android.Progress
import org.syncloud.android.tasks.AsyncResult.Companion.exception
import org.syncloud.android.tasks.AsyncResult.Companion.value

class ProgressAsyncTask<TParams, TResult> @JvmOverloads constructor(
    private var progress: Progress? = null,
    private var title: String? = null,
    private var work: Work<TParams, TResult>? = null
) : AsyncTask<TParams, Void, AsyncResult<TResult>>() {
    interface Success<TResult> {
        fun run(result: TResult?)
    }

    interface Completed<TResult> {
        fun run(result: AsyncResult<TResult>?)
    }

    interface Work<TParams, TResult> {
        fun run(vararg args: TParams): TResult
    }

    private var errorMessage: String? = null
    private var success: Success<TResult>? = null
    private var completed: Completed<TResult>? = null

    override fun doInBackground(vararg args: TParams): AsyncResult<TResult>? {
        return if (work != null) {
            try {
                value(work!!.run(*args))
            } catch (th: Throwable) {
                exception(th)
            }
        } else {
            null
        }
    }

    fun setErrorMessage(message: String): ProgressAsyncTask<TParams, TResult> {
        errorMessage = message
        return this
    }

    fun setTitle(title: String): ProgressAsyncTask<TParams, TResult> {
        this.title = title
        return this
    }

    fun setProgress(progress: Progress): ProgressAsyncTask<TParams, TResult> {
        this.progress = progress
        return this
    }

    fun onSuccess(success: Success<TResult>): ProgressAsyncTask<TParams, TResult> {
        this.success = success
        return this
    }

    fun onCompleted(completed: Completed<TResult>): ProgressAsyncTask<TParams, TResult> {
        this.completed = completed
        return this
    }

    fun doWork(work: Work<TParams, TResult>): ProgressAsyncTask<TParams, TResult> {
        this.work = work
        return this
    }

    override fun onPreExecute() {
        if (progress != null) {
            progress!!.start()
            if (title != null) progress!!.title(title)
        }
    }

    override fun onPostExecute(result: AsyncResult<TResult>?) {
        if (progress != null) {
            if (result != null && !result.hasValue() && errorMessage != null) progress!!.error(
                errorMessage
            ) else progress!!.stop()
        }
        if (result == null) success!!.run(null) else if (result.hasValue() && success != null) success!!.run(
            result.getValue()
        )
        if (completed != null) completed!!.run(result)
    }

    override fun onCancelled() {
        if (progress != null) {
            progress!!.stop()
        }
    }

    companion object {
        fun execute(runnable: Runnable) {
            ProgressAsyncTask<Void, Unit>()
                .doWork(object : Work<Void, Unit> {
                    override fun run(vararg args: Void): Unit {
                        return runnable.run()
                    }
                })
                .execute()
        }
    }
}